src/java.base/share/classes/jdk/internal/module/ModuleBootstrap.java

```diff
@@ -469,15 +468,27 @@ private static ModuleLayer boot2() {
                 limitedFinder = new SafeModuleFinder(finder);
         }
 
+        // If -Xshare:dump and mainModule are specified, check if the mainModule
+        // is in the runtime image and not in the upgrade module path. If so,
+        // set canArchive to true so that the module graph can be archived.
+        if (CDS.isDumpingStaticArchive() && mainModule != null) {
+            ModuleReference mainModuleRef = systemModuleFinder.find(mainModule).orElse(null);
+            if (mainModuleRef != null) {
+                URI mainModuleLocation = mainModuleRef.location().orElse(null);
+                if (mainModuleLocation != null && mainModuleLocation.getScheme().equalsIgnoreCase("jrt")) {
+                    canArchive = true;
+                }
+            }
+        }
+
         // Archive module graph and boot layer can be archived at CDS dump time.
-        // Only allow the unnamed module case for now.
-        if (canArchive && (mainModule == null)) {
+        if (canArchive) {
             ArchivedModuleGraph.archive(hasSplitPackages,
                                         hasIncubatorModules,
                                         systemModuleFinder,
                                         cf,
                                         clf);
-            if (!hasSplitPackages && !hasIncubatorModules) {
+            if (!hasSplitPackages) {
```

Comment:
Did you mean to drop the hasIncubatorModules check? Incubator modules are not resolved by default.

Reply:
Yes, because of the following code further up in the same method:
```
            if (!haveModulePath && addModules.isEmpty() && limitModules.isEmpty()) {
                systemModules = SystemModuleFinders.systemModules(mainModule);
                if (systemModules != null && !isPatched) {
                    needResolution = (traceOutput != null);
                    canArchive = true;
                }
            }
            if (systemModules == null) {
                // all system modules are observable
                systemModules = SystemModuleFinders.allSystemModules();
            }
```
`SystemModuleFinders.systemModules(mainModule)` returns null if `mainModule` is not null (I'm not sure why).
Then, the `SystemModuleFinders.allSystemModules()` returns the generated class `SystemModules$all` with the following method:
```
    public boolean hasIncubatorModules() {
        return true;
    }
```

Reply:
> Yes, because of the following code further up in the same method:

I think what you are actually doing here is supporting archiving of the boot layer when the main module transitively depends on an incubator module. We might have to add more tests in this area as it doesn't look like this scenario is tested, at least not with CDS.

Reply:
> > Yes, because of the following code further up in the same method:
> 
> I think what you are actually doing here is supporting archiving of the boot layer when the main module transitively depends on an incubator module. We might have to add more tests in this area as it doesn't look like this scenario is tested, at least not with CDS.

Just to add to this. The effect of this change is that a warning won't be emitted if you archive with a boot layer that contains a main module and an incubator module, e.g. main requires jdk.incubator.vector. In other words, `java -Xshare:off -m main` will print the expected warning, `java -Xshare:off -m main` will not.

Reply:
Another effect of the change to note is that incubator modules will now resolved by default when the initial module is the unnamed module. More explicitly, if you jlink to create a run-time image that contains a main module and an incubator module then `java -Xshare:off -version` will not resolve the incubator moduls (correct behaviour) but `java -Xshare:on` will resolve incubator modules (wrong behavior).

Reply:
I've pushed another update with the following changes:

- in the VM code, skip archiving full module graph if there's an incubator module by checking if the ArchivedBootLayer::archivedBootLayer is available;
- included your suggested code changes in ModuleBootstrap.java;
- added a test scenario with an incubator module.

Reply:
> I've pushed another update with the following changes:
> 
> * in the VM code, skip archiving full module graph if there's an incubator module by checking if the ArchivedBootLayer::archivedBootLayer is available;

I think we still have the issue when the archive is created with a main module (meaning a named module) but you run with the class path as the initial module, e.g. java -version. Is the update missing the changes to ArchivedModuleGraph as I think that will need to check that the main module matches the archive.

Reply:
I reran the script you sent me few days ago and got the expected results with the latest changes.
The checking of the main module name matches between dump time and runtime is performed in the VM code. If an archive (even the default CDS archive classes.jsa) is created with the -m option, running with just `java -version` will disable the optimized module handling.
```
[0.136s][info ][cds        ] _archived_main_module_name m
[0.136s][info ][cds        ] Module m specified during dump time but not during runtime
[0.136s][info ][cds        ] Disabling optimized module handling
[0.136s][info ][cds        ] optimized module handling: disabled
[0.136s][info ][cds        ] full module graph: disabled
```

Reply:
> I reran the script you sent me few days ago and got the expected results with the latest changes. The checking of the main module name matches between dump time and runtime is performed in the VM code. If an archive (even the default CDS archive classes.jsa) is created with the -m option, running with just `java -version` will disable the optimized module handling.
> 
> ```
> [0.136s][info ][cds        ] _archived_main_module_name m
> [0.136s][info ][cds        ] Module m specified during dump time but not during runtime
> [0.136s][info ][cds        ] Disabling optimized module handling
> [0.136s][info ][cds        ] optimized module handling: disabled
> [0.136s][info ][cds        ] full module graph: disabled
> ```

Can you check that ArchivedModuleGraph.archivedModuleGraph is null when disabled? When I tried it, it wasn't null so ModuleBootstrap has a module graph containing the modules for m.

Reply:
The `ArchivedModuleGraph.java` wasn't changed. So if `-m` is not specified, the `archivedModuleGraph` is non-null; if `-m` is specified, the `archivedModuleGraph` is null.
So running `java -version`, the archivedModuleGraph is non-null, but the module m won't be loaded from the archive. I'm seeing the following log output:

`[0.220s][debug][module     ] define_module(): creation of module: m, version: null, location: jrt:/m, loader data: 0x00007f201819dd60 for instance a 'jdk/internal/loader/ClassLoaders$AppClassLoader'{0x000000011f010500}, package #: 1`

Reply:
> The `ArchivedModuleGraph.java` wasn't changed. So if `-m` is not specified, the `archivedModuleGraph` is non-null; if `-m` is specified, the `archivedModuleGraph` is null. So running `java -version`, the archivedModuleGraph is non-null, but the module m won't be loaded from the archive.

Right, but ArchivedModuleGraph provides the initial value for hasIncubatorModules, hasSplitPackages, ... This is why `java -version` confusingly prints a warning that an incubator module has been resolved. We've had archiving of the module graph (configuration) for a long time but it's only been for the case where the initial module is the unnamed module. I think archiving of the boot layer for the case where the initial module is a named module means we have to re-visit ArchivedModuleGraph. I think ArchivedModuleGraph.get has to return null when the archive doesn't match the mainModule. Part of it wonder if we should remove ArchivedModuleGraph, support archiving of the boot layer only.

Reply:
I've added the following field in `ArchivedModuleGraph` so that the `get(String mainModuleName)` will check the `mainModule` before returning `archivedModuleGraph`.
`private static String mainModule;`
The `mainModule` field is also added in `heapShared.cpp` so that it is stored in the archive.

Reply:
> I've added the following field in `ArchivedModuleGraph` so that the `get(String mainModuleName)` will check the `mainModule` before returning `archivedModuleGraph`. `private static String mainModule;` The `mainModule` field is also added in `heapShared.cpp` so that it is stored in the archive.

Thanks, it looks correctly now.

One small question. At ModuleBootstrap L235 we set canArchive as it's okay to archive under specific restrictions. For completeness, shouldn't this set canArchive to CDS.isDumpingStaticArchive? I don't think it matters right now but might be confusing to have canArchive be true when not dumping.

Reply:
> Thanks, it looks correctly now.

Thanks!

> 
> One small question. At ModuleBootstrap L235 we set canArchive as it's okay to archive under specific restrictions. For completeness, shouldn't this set canArchive to CDS.isDumpingStaticArchive? I don't think it matters right now but might be confusing to have canArchive be true when not dumping.

Do you mean L230?
```
 226             if (!haveModulePath && addModules.isEmpty() && limitModules.isEmpty()) {
 227                 systemModules = SystemModuleFinders.systemModules(mainModule);
 228                 if (systemModules != null && !isPatched) {
 229                     needResolution = (traceOutput != null);
 230                     canArchive = true;
 231                 }
 232             }
```
Do you prefer the `canArchive` setting be inside `if (CDS.isDumpingStaticArchive())` like the following?
```
    if (CDS.isDumpingStaticArchive())
        canArchive = true;
```

Reply:
> Do you prefer the `canArchive` setting be inside `if (CDS.isDumpingStaticArchive())` like the following?
> 
> ```
>     if (CDS.isDumpingStaticArchive())
>         canArchive = true;
> ```

That's fine. It has been harmless to have canArchive be true when not dumping but with the other changes in this PR then I think it's clearer to only set it when dumping. So I think your changes are good.